core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -289,19 +311,24 @@ int readVersionHint() {
       return Integer.parseInt(in.readLine().replace("\n", ""));
 
     } catch (Exception e) {
-      LOG.warn("Error reading version hint file {}", versionHintFile, e);
+      LOG.debug("Error reading version hint file {}", versionHintFile, e);
       try {
-        if (getMetadataFile(1) != null) {
-          // We just assume corrupted metadata and start to read from the first version file
-          return 1;
+        // List the metadata directory to find the version files, and try to recover the max available version
+        FileStatus[] files = fs.listStatus(metadataRoot(), name -> VERSION_PATTERN.matcher(name.getName()).matches());
+        int maxVersion = 0;
+
+        for (FileStatus file : files) {
+          version = version(file.getPath().getName());
```

Comment:
I didn't catch this at first, but this method should not have a side effect of setting `version`. It should return version and something else should set it.

Also, when we set instance fields, we use the prefix `this.` so that it is clear that an instance field and not a local variable is set. That's why I missed this in my first round of review.

Reply:
I was so focused on not finding a way to replicate the failures on my side, that I missed this.
Big KUDOS for catching this! :)
I hope fixing this will fix the tests as well (Still not sure why they were passing on my side 😢)

Reply:
This fixed the tests. Still not sure why my tests did not repro the case 😢